src/hotspot/cpu/aarch64/templateInterpreterGenerator_aarch64.cpp

```diff
@@ -1978,21 +1978,37 @@ address TemplateInterpreterGenerator::generate_trace_code(TosState state) {
 }
 
 void TemplateInterpreterGenerator::count_bytecode() {
-  Register rscratch3 = r0;
-  __ push(rscratch1);
-  __ push(rscratch2);
-  __ push(rscratch3);
-  __ mov(rscratch3, (address) &BytecodeCounter::_counter_value);
-  __ atomic_add(noreg, 1, rscratch3);
-  __ pop(rscratch3);
-  __ pop(rscratch2);
-  __ pop(rscratch1);
+  __ mov(r10, (address) &BytecodeCounter::_counter_value);
+  __ atomic_addw(noreg, 1, r10);
 }
 
-void TemplateInterpreterGenerator::histogram_bytecode(Template* t) { ; }
-
-void TemplateInterpreterGenerator::histogram_bytecode_pair(Template* t) { ; }
+void TemplateInterpreterGenerator::histogram_bytecode(Template* t) {
+  __ mov(r10, (address) &BytecodeHistogram::_counters[t->bytecode()]);
+  __ atomic_addw(noreg, 1, r10);
+}
 
+void TemplateInterpreterGenerator::histogram_bytecode_pair(Template* t) {
+  // Calculate new index for counter:
+  //   _index = (_index >> log2_number_of_codes) |
+  //            (bytecode << log2_number_of_codes);
+  Register index_addr = r10;
+  Register index = r11;
+  Label L;
+  __ mov(index_addr, (address) &BytecodePairHistogram::_index);
+  __ mov(rscratch1,
+         ((int)t->bytecode()) << BytecodePairHistogram::log2_number_of_codes);
+  __ atomic_orrw(index_addr, rscratch1, index,
+                 /* tmp */ rscratch2,
+                 /* kind */ Assembler::LSR,
+                 /* shift */ BytecodePairHistogram::log2_number_of_codes);
+
```

Comment:
I've had another look at this. `_index` is a two-element queue of bytecodes, but it is shared between all threads. If two threads access `_index` racily the result will be invalid, regardless of whether the OR into memory is atomic. This will make this PR much simpler.
None of the other ports access `_index` atomically, and should need we. On the other hand, bumping the bytecode counter atomically is fine.
We could make a thread-local `_index`, but it's too much code to be worthwhile.

Reply:
Thanks for pointing this out. Updated.